appkit/src/test/scala/org/ergoplatform/appkit/AppkitProvingInterpreterSpec.scala

```diff
@@ -0,0 +1,189 @@
+    val stateContext = txB.createErgoLikeStateContext
+    val changeAddress = prover.getAddress.getErgoAddress
+
+    val boxesToSpend = inputs
+      .map(b => new InputBoxImpl(b))
+      .map(b => ExtendedInputBox(b.getErgoBox, b.getExtension))
+      .convertTo[util.List[ExtendedInputBox]]
+    val boxesToSpendSeq = JavaHelpers.toIndexedSeq(boxesToSpend)
+    val tx = new UnsignedErgoLikeTransaction(
+      inputs = boxesToSpendSeq.map(_.toUnsignedInput),
+      dataInputs = IndexedSeq(),
+      outputCandidates = outputs
+    )
+    val unsigned = new UnsignedTransactionImpl(
+      tx, boxesToSpend, new util.ArrayList[ErgoBox](), changeAddress, stateContext,
+      ctx.asInstanceOf[BlockchainContextImpl], tokensToBurn.convertTo[JList[ErgoToken]])
+    unsigned
+  }
+
+  property("rejecting transaction with unbalanced tokens") {
+    val ergoClient = createMockedErgoClient(MockData(Nil, Nil))
+    ergoClient.execute { ctx: BlockchainContext =>
+      val prover = ctx.newProverBuilder()
+        .withMnemonic(mnemonic, SecretString.empty())
+        .build()
+      val tree1 = ErgoScriptPredef.TrueProp(ergoTreeHeaderInTests)
+      val tree2 = ErgoScriptPredef.FalseProp(ergoTreeHeaderInTests)
+      val token1 = (ErgoAlgos.hash("id1"), 10L)
+      val token2 = (ErgoAlgos.hash("id2"), 20L)
+      val ergoToken1 = Iso.isoErgoTokenToPair.from(token1)
+      val ergoToken2 = Iso.isoErgoTokenToPair.from(token2)
+
+      val input1 = createBox(oneErg + Parameters.MinFee, tree1, additionalTokens = Seq(token1))
+      val input2 = createBox(oneErg, tree2, additionalTokens = Seq(token2))
+
+      // successful reduction with balanced tokens
+      {
+        val ops = createBoxOps(ctx, prover, IndexedSeq(input1, input2))
+        val tokens = new util.ArrayList[ErgoToken]()
+        tokens.add(ergoToken1)
+        tokens.add(ergoToken2)
+        val unsigned = ops
+          .withTokensToSpend(tokens)
+          .putToContractTxUnsigned(address.toErgoContract)
+        val reduced = prover.reduce(unsigned, 0)
+        reduced.getInputBoxesIds.size() shouldBe 2
+        reduced.getOutputs.size() shouldBe 2  // output + feeOut
+      }
+
+      // Transaction tries to burn tokens when no burning was requested
+      {
+        val output1 = createBox(oneErg * 2 + Parameters.MinFee, tree1, additionalTokens = Seq(token1))
+        val unsigned = createUnsignedTransaction(ctx, prover, IndexedSeq(input1, input2), IndexedSeq(output1), IndexedSeq.empty)
+        assertExceptionThrown(
+          prover.reduce(unsigned, 0),
+          {
+            case e: TokenBalanceException =>
+              val cond1 = exceptionLike[TokenBalanceException]("Transaction tries to burn tokens when no burning was requested")
+              cond1(e) && e.tokensDiff.exists(t => t == (Colls.fromArray(token2._1), -token2._2))
+            case _ => false
+          }
+        )
+      }
+
+      // Transaction tries to burn tokens when no burning was requested
+      {
```

Comment:
Good to have here would be a test that test with the following inputs and outputs:
Inputs:
Box1: Token1, Amount1
Box2: Token1. Amount2

Outputs:
Box1: Token1, Amount1

So that we have a test that checks if grouping a token is implemented correct.

Reply:
done